ddtrace/internal/module.py

```diff
@@ -0,0 +1,482 @@
+            if loader and not isinstance(loader, _ImportHookChainedLoader):
+                return _ImportHookChainedLoader(loader, self.after_import)
+
+        finally:
+            self._finding.remove(fullname)
+
+        return None
+
+    def find_spec(self, fullname, path=None, target=None):
+        # type: (str, Optional[str], Optional[ModuleType]) -> Optional[ModuleSpec]
+        if fullname in self._finding:
+            return None
+
+        self._finding.add(fullname)
+
+        try:
+            spec = find_spec(fullname)
+            if spec is None:
+                return None
+
+            loader = getattr(spec, "loader", None)
+
+            if loader and not isinstance(loader, _ImportHookChainedLoader):
+                spec.loader = _ImportHookChainedLoader(loader, self.after_import)
+
+            return spec
+
+        finally:
+            self._finding.remove(fullname)
+
+    @classmethod
+    def register_origin_hook(cls, origin, hook, arg):
+        # type: (str, HookType, Any) -> None
+        """Register a hook to be called when the module with the given origin is
+        imported.
+
+        The hook will be called with the module object and the given argument.
+        """
+        cls._check_installed()
+
+        # DEV: Under the hypothesis that this is only ever called by the probe
+        # poller thread, there are no further actions to take. Should this ever
+        # change, then thread-safety might become a concern.
+        path = _resolve(origin)
+        if path is None:
+            raise ValueError("Cannot resolve module origin %s" % origin)
+
+        log.debug("Registering hook '%r' on path '%s'", hook, path)
+        instance = cast(ModuleWatchdog, cls._instance)
+        instance._hook_map[path].append((hook, arg))
+        try:
+            module = instance._origin_map[path]
+        except KeyError:
+            # The module is not loaded yet. Nothing more we can do.
+            return
+
+        # The module was already imported so we invoke the hook straight-away
+        log.debug("Calling hook '%r' on already imported module '%s'", hook, module.__name__)
+        hook(module, arg)
+
+    @classmethod
+    def unregister_origin_hook(cls, origin, arg):
+        # type: (str, Any) -> None
+        """Unregister the hook registered with the given module origin and
+        argument.
+        """
+        cls._check_installed()
+
+        path = _resolve(origin)
+        if path is None:
+            raise ValueError("Module origin %s cannot be resolved", origin)
+
+        instance = cast(ModuleWatchdog, cls._instance)
+        if path not in instance._hook_map:
+            raise ValueError("No hooks registered for origin %s" % origin)
+
+        for i, (_, a) in enumerate(instance._hook_map[path]):
+            if isinstance(a, list) and arg in a:
+                # DEV: This comes from the knowledge that sometimes the argument
+                # of the hook can be a list.
```

Comment:
So we can do

```python
register_origin_hook(mod, hook, [1, 2, 3])
```

and then 

```python
unregister_origin_hook(mod, hook, 3)
```
?

We should probably document this usage in the docstrings. Also I don't think we have tests for this logic.

Reply:
Yeah that's also task-specific code and it's going away with the removal of the `arg` 🧹